src/backend/columnar/columnar_metadata.c

```diff
@@ -953,35 +1040,123 @@ GetHighestUsedAddressAndId(uint64 storageId,
 
 
 /*
- * ReserveStripe reserves and stripe of given size for the given relation,
+ * ReserveEmptyStripe reserves an empty stripe for given relation
  * and inserts it into columnar.stripe. It is guaranteed that concurrent
  * writes won't overwrite the returned stripe.
  */
-StripeMetadata
-ReserveStripe(Relation rel, uint64 sizeBytes,
-			  uint64 rowCount, uint64 columnCount,
-			  uint64 chunkCount, uint64 chunkGroupRowCount,
-			  uint64 stripeFirstRowNumber)
+EmptyStripeReservation *
+ReserveEmptyStripe(Relation rel, uint64 columnCount, uint64 chunkGroupRowCount,
+				   uint64 stripeRowCount)
 {
-	StripeMetadata stripe = { 0 };
+	EmptyStripeReservation *stripeReservation = palloc0(sizeof(EmptyStripeReservation));
 
 	uint64 storageId = ColumnarStorageGetStorageId(rel, false);
 
-	uint64 stripeId = ColumnarStorageReserveStripe(rel);
+	/* do stripe id and first row number reservations atomically */
+	LockRelationForExtension(rel, ExclusiveLock);
```

Comment:
This should be handled by columnar_storage, right?

Reply:
Right, then should I add a new function to columnar_storage that reserves both stripe & first_row_number using the lower-level functions ?

Reply:
Is there a reason those need to be done automically? I'm fine with adding a function like that, it makes sense, I'm just wondering if it's a requirement.

Reply:
Actually there is no reason for doing both atomically, just wanted to be consistent.

Does it still makes sense if I add a function for that to `columnar_storage.c`, or do you think it wouldn't worth as it is not a requirement ?